File: Programmers/연습문제/Level 1/크레인 인형뽑기 게임/Solution.java
import java.util.Stack;

class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        Stack<Integer> bucket = new Stack<>();
        for (int move : moves) {
            move -= 1;
            for (int depth = 0; depth < board.length; depth++) {
                int doll = board[depth][move];
                if (doll != 0) {
                    board[depth][move] = 0;
                    if (!bucket.isEmpty() && bucket.peek() == doll) {
                        bucket.pop();
                        answer += 2;
                    } else {
                        bucket.push(doll);
                    }
                    break;
                }
            }
        }
        return answer;
    }
}

File: Programmers/연습문제/Level 1/크레인 인형뽑기 게임/Solution.py
def solution(board, moves):
    answer = 0
    bucket = []
    for move in moves:
        move -= 1
        for depth in range(len(board)):
            doll = board[depth][move]
            if doll != 0:
                bucket.append(doll)
                board[depth][move] = 0
                if len(bucket) >= 2 and bucket[-1] == bucket[-2]:
                    bucket.pop()
                    bucket.pop()
                    answer += 2
                break
    return answer


if __name__ == "__main__":
    board = [[0,0,0,0,0],[0,0,1,0,3],[0,2,5,0,1],[4,2,4,4,2],[3,5,1,3,1]]
    moves = [1,5,3,5,1,2,1,4]
    print(solution(board, moves)) # 4